Use deprecated key alias if already present.

morganlutz opened the pull request:
This change affects pre-existing data. New logins should encrypt/decrypt data with the new key alias

dsuresh-ap committed:
Update README with callback threading note

Update README to note that callback methods might not be run on the main thread. Related to issue filed here: https://github.com/auth0/Auth0.Android/issues/316.

morganlutz committed:
Use deprecated key alias if already present.

+ This change only affects pre-existing data. New logins should encrypt/decrypt data with the new key alias

bryfox commented on auth0/src/main/java/com/auth0/android/authentication/storage/CryptoUtil.java:
```diff
@@ -93,9 +97,18 @@ public CryptoUtil(@NonNull Context context, @NonNull Storage storage, @NonNull S
-                KeyStore.PrivateKeyEntry existingKey = getKeyEntryCompat(keyStore);
+                KeyStore.PrivateKeyEntry existingKey = getKeyEntryCompat(keyStore, alias);
                 if (existingKey != null) {
                     return existingKey;
```
This is where, as long as we have a valid RSA key, we skip generation of a new one. I don't see any key rotation happening; the old one appears to be deleted only if it is corrupted. If our intent is to phase this out quickly, I think we need to implement the rotation (or be OK with requiring a new log-in).

bryfox replied:
Same basic idea for AES, I think.

bryfox commented on auth0/src/main/java/com/auth0/android/authentication/storage/CryptoUtil.java:
```diff
@@ -400,6 +419,9 @@ private void deleteAESKeys() {
+                encodedIV = storage.retrieveString(DEPRECATED_KEY_IV_ALIAS);
+            }
             if (TextUtils.isEmpty(encodedIV)) {
                 //AES key was JUST generated. If anything existed before, should be encrypted again first.
```
Does this imply some consistency issue with storage?

bryfox commented on auth0/src/main/java/com/auth0/android/authentication/storage/CryptoUtil.java:
```diff
@@ -400,6 +419,9 @@ private void deleteAESKeys() {
             SecretKey key = new SecretKeySpec(getAESKey(), ALGORITHM_AES);
             Cipher cipher = Cipher.getInstance(AES_TRANSFORMATION);
             String encodedIV = storage.retrieveString(KEY_IV_ALIAS);
+            if (TextUtils.isEmpty(encodedIV)) {
```
If there's any possibility of the new alias being in use but the IV value being empty (seems unlikely, but then again the comment below implies it's possible), then this could find an old value?

I suppose that would just result in the same result (forced logout), as long as the KEY_IV_ALIAS read succeeded the second time. Might be something to consider if we need rotation.

bryfox reviewed (commented):
I think this seems good, but may not have the desired effect of migrating users to the new keys.

lbalmaceda committed:
add test cases for migration scenario

lbalmaceda committed:
fix secure credential manager key alias migration

lbalmaceda committed:
Add thread switch note

lbalmaceda committed:
Merge pull request #322 from dsuresh-ap/patch-1

Update README with callback threading note

lbalmaceda committed:
Merge branch 'master' into patch-sec-cm

lbalmaceda committed:
Merge pull request #325 from auth0/patch-sec-cm

Patch Key alias migration for Secure Credentials Manager

lbalmaceda committed:
Release 1.24.1

lbalmaceda committed:
Merge pull request #326 from auth0/perform-release

Release 1.24.1

lbalmaceda committed:
Update CHANGELOG.md

lbalmaceda committed:
Update note title

lbalmaceda committed:
Add failing test for custom issuer validation

lbalmaceda committed:
Add logic to accept custom IDTV issuer

lbalmaceda committed:
Add tests for null custom IDTV issuer

lbalmaceda committed:
Merge pull request #327 from auth0/lbalmaceda-readme-patch-1

Add a note to the CHANGELOG file about the recent release

lbalmaceda committed:
add note to the readme about custom issuer

lbalmaceda committed:
update readme

lbalmaceda committed:
add getter for error of type "verification required"

lbalmaceda committed:
add snippet and note about bot protection to the readme

lbalmaceda committed:
Use the same connection value on webauth sample

lbalmaceda committed:
Fix snippet quotes

lbalmaceda committed:
Merge pull request #328 from auth0/custom-issuer

Support use of Custom Issuer for ID Token verification [SDK-1910]

lbalmaceda committed:
Add screen_hint guidance to the readme

lbalmaceda committed:
Update README.md

Co-authored-by: Rita Zerrizuela <zeta@widcket.com>

lbalmaceda committed:
Merge branch 'master' into feat-verify-request

lbalmaceda committed:
Apply suggestions from code review

Co-authored-by: Rita Zerrizuela <zeta@widcket.com>

lbalmaceda committed:
Merge pull request #329 from auth0/feat-verify-request

Add Bot Protection support

lbalmaceda committed:
Release 1.25.0

Widcket committed:
Merge pull request #331 from auth0/perform-release

Release 1.25.0

lbalmaceda committed:
add support for android 11 new privacy settings

nelsonmaia committed:
Adding support to override the browser packages to be used during Login / Logout

Adding support to override the browser packages to be used during Login / Logout

lbalmaceda committed:
Merge pull request #335 from auth0/fix-queries

Add support for Android 11 new privacy settings

fossabot committed:
Add license scan report and status

Signed off by: fossabot <badges@fossa.com>

lbalmaceda committed:
Merge pull request #336 from fossabot/add-license-scan-badge

Add license scan report and status

nelsonmaia committed:
Adding support to search all the browsers in the device

lbalmaceda committed:
Improve compatibility with Kotlin and run Lint on CI (#337)

* run linter and improve interoperability

* improve the CI script to run lint checks

lbalmaceda committed:
Release 1.26.0 (#339)

lbalmaceda committed:
Add update notice (#340)

This notice was included in the PR but missing from the changelog. With this, I hope to give more visibility.

jsalinaspolo committed:
Official Gradle Wrapper Validation Action

See: https://github.com/gradle/wrapper-validation-action

lbalmaceda committed:
update non-null annotations to fix NPE on kotlin (#344)

lbalmaceda committed:
Release 1.26.1

jsalinaspolo committed:
Rename gh-action name

lbalmaceda committed:
refactor to use extensible BrowserPicker filtering

lbalmaceda committed:
Fix tests for changed logic

lbalmaceda committed:
Replace deprecated import with the alternative

lbalmaceda committed:
add tests for the BrowserPicker class

lbalmaceda committed:
add coverage for introduced use cases and javadocs

lbalmaceda committed:
add CompatibleBrowser tests

lbalmaceda committed:
fix flag passed when querying browser apps for post marshmallow

lbalmaceda committed:
Release 1.27.0

quibi-jlk committed:
Catch a gson JsonIOException when parsing SimpleRequest response.

IOExceptions are already caught, but it turns out in some cases, gson actually wraps underlying IOExceptions in JsonIOExceptions (which does not extent IOException).

Example stack trace:
```
Fatal Exception: com.google.gson.JsonIOException: java.net.SocketTimeoutException: timeout
   at com.google.gson.internal.Streams.parse(Streams.java:62)
   at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:65)
   at com.google.gson.TypeAdapter.fromJson(TypeAdapter.java:260)
   at com.auth0.android.request.internal.SimpleRequest.onResponse(SimpleRequest.java:76)
   at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:177)
   at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
   at java.lang.Thread.run(Thread.java:764)

Caused by java.net.SocketTimeoutException: timeout
   at okio.SocketAsyncTimeout.newTimeoutException(SocketAsyncTimeout.java:143)
   at okio.AsyncTimeout.access$newTimeoutException(AsyncTimeout.java:162)
   at okio.AsyncTimeout.withTimeout(AsyncTimeout.java:154)
   at okio.AsyncTimeout$source$1.read(AsyncTimeout.java:340)
   at okio.internal.RealBufferedSourceKt.commonRead(RealBufferedSourceKt.java:39)
   at okio.RealBufferedSource.read(RealBufferedSource.java:188)
   at com.squareup.okhttp.internal.http.Http1xStream$ChunkedSource.read(Http1xStream.java:439)
   at okio.internal.RealBufferedSourceKt.commonRead(RealBufferedSourceKt.java:39)
   at okio.RealBufferedSource.read(RealBufferedSource.java:188)
   at okio.internal.RealBufferedSourceKt.commonExhausted(RealBufferedSourceKt.java:49)
   at okio.RealBufferedSource.exhausted(RealBufferedSource.java:198)
   at okio.InflaterSource.refill(InflaterSource.java:112)
   at okio.InflaterSource.readOrInflate(InflaterSource.java:76)
   at okio.InflaterSource.read(InflaterSource.java:49)
   at okio.GzipSource.read(GzipSource.java:69)
   at okio.RealBufferedSource$inputStream$1.read(RealBufferedSource.java:158)
   at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:288)
   at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:351)
   at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:180)
   at java.io.InputStreamReader.read(InputStreamReader.java:184)
   at com.google.gson.stream.JsonReader.fillBuffer(JsonReader.java:1291)
   at com.google.gson.stream.JsonReader.nextNonWhitespace(JsonReader.java:1329)
   at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:550)
   at com.google.gson.stream.JsonReader.peek(JsonReader.java:426)
   at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:700)
   at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:723)
   at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:698)
   at com.google.gson.internal.Streams.parse(Streams.java:48)
   at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:65)
   at com.google.gson.TypeAdapter.fromJson(TypeAdapter.java:260)
   at com.auth0.android.request.internal.SimpleRequest.onResponse(SimpleRequest.java:76)
   at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:177)
   at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
   at java.lang.Thread.run(Thread.java:764)
```

quibi-jlk committed:
Switch to a JsonParseException.

lbalmaceda committed:
catch JsonParseException when parsing sync requests

lbalmaceda committed:
cover SimpleRequest with unit tests

lbalmaceda committed:
fix incomplete verify statement

lbalmaceda committed:
improve tests, match correct exception class

lbalmaceda committed:
accept a custom clock instance in the credentials managers

lbalmaceda committed:
Release 1.28.0

lbalmaceda committed:
Merge pull request #361 from auth0/perform-release

Release 1.28.0
